core/arch/arm/mm/core_mmu.c

```diff
@@ -2474,3 +2481,98 @@ void core_mmu_init_ta_ram(void)
 	tee_mm_init(&tee_mm_sec_ddr, ps, pe, CORE_MMU_USER_CODE_SHIFT,
 		    TEE_MM_POOL_NO_FLAGS);
 }
+
+#ifdef CFG_NS_RTI_CHECK
+void *core_mmu_map_rti_check(paddr_t pa, size_t len, size_t *mapped_len)
+{
+	uint32_t new_attr = TEE_MATTR_VALID_BLOCK | TEE_MATTR_PR |
+			    TEE_MATTR_GLOBAL |
+			    (TEE_MATTR_CACHE_CACHED << TEE_MATTR_CACHE_SHIFT);
+	struct core_mmu_table_info tbl_info = { };
+	struct tee_mmap_region *map = NULL;
+	size_t old_granule = 0;
+	vaddr_t vabase = 0;
+	uint32_t attr = 0;
+	size_t idx = 0;
+
+	/* Refuse to map memory which isn't know to be non-secure */
+	if (!core_pbuf_is(CORE_MEM_NON_SEC, pa, len))
+		return NULL;
+
+	map = find_map_by_type(MEM_AREA_RTI_CHECK_VASPACE);
+	if (!map)
+		panic("MEM_AREA_RTI_CHECK_VASPACE not found");
+	vabase = map->va;
+
+	if (!core_mmu_find_table(NULL, vabase, CORE_MMU_PGDIR_LEVEL - 1,
+				 &tbl_info))
+		panic("Can't find pgdir entry");
+
+	idx = core_mmu_va2idx(&tbl_info, vabase);
+	core_mmu_get_entry(&tbl_info, idx, NULL, &attr);
+	if (attr & TEE_MATTR_TABLE)
+		old_granule = SMALL_PAGE_SIZE;
+	else
+		old_granule = CORE_MMU_PGDIR_SIZE;
+
+	/*
+	 * Invalidate the map and matching TLB entries before setting the
+	 * new map.
+	 */
+	if (attr & (TEE_MATTR_PR | TEE_MATTR_TABLE)) {
+		core_mmu_set_entry(&tbl_info, idx, 0, 0);
+		tlbi_mva_range(vabase, CORE_MMU_PGDIR_SIZE, old_granule);
+	}
+	if (!len)
+		return NULL;
+	/* We cannot map anything smaller than a page */
+	if ((pa | len) & SMALL_PAGE_MASK)
+		return NULL;
+
```

Comment:
Those two could be moved at the start

Reply:
I'd like a `NULL` return to mean that no memory is mapped any longer. That way there's no need for extra calls to make sure nothing is mapped.